From ff72fb041b16945ace6b0597216a91eba601acb8 Mon Sep 17 00:00:00 2001
From: Narendhar Reddy <nare@qti.qualcomm.com>
Date: Thu, 10 Jul 2025 15:06:26 +0530
Subject: [PATCH] QSDK:DP:ath12k: fix memory leak in ppeds.

Currently in rmmod/insmod usecase, memory allocated
for tx_status and ppeds_rx_idx is not getting freed, leading to
out of memory issue during continuous rmmod/insmod.

Fix the leak by propely freeing the memory.

Fixes : I56018d1a5b
patch dependency: 0337-01-QSDK-DP-wifi-Add-support-for-PPE-DS-Direct-switch-fe.patch
Signed-off-by: Narendhar Reddy <nare@qti.qualcomm.com>
---
 drivers/net/wireless/ath/ath12k/ppe.c | 3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)

diff --git a/drivers/net/wireless/ath/ath12k/ppe.c b/drivers/net/wireless/ath/ath12k/ppe.c
index 24bd4a1..a8b680e 100644
--- a/drivers/net/wireless/ath/ath12k/ppe.c
+++ b/drivers/net/wireless/ath/ath12k/ppe.c
@@ -1283,7 +1283,7 @@ int ath12k_ppeds_attach(struct ath12k_base *ab)
 	*abptr = ab;
 	ath12k_info(ab, "PPEDS attach success\n");
 
-	for (i = 0; i < ATH12K_MAX_SOCS; i++) {
+	for (i = 0; i < ab->ag->num_hw; i++) {
 		ab->dp->ppe.ppeds_rx_idx[i] = kzalloc((sizeof(u16) * PPE_DS_TXCMPL_DEF_BUDGET),
 						      GFP_ATOMIC);
 		if (!ab->dp->ppe.ppeds_rx_idx[i]) {
@@ -1716,6 +1716,7 @@ void ath12k_dp_srng_ppeds_cleanup(struct ath12k_base *ab)
 
 	ath12k_dp_srng_cleanup(ab, &dp->ppe.ppe2tcl_ring);
 	ath12k_dp_srng_cleanup(ab, &dp->ppe.reo2ppe_ring);
+	kfree(dp->ppe.ppeds_comp_ring.tx_status);
 	ath12k_dp_srng_cleanup(ab, &dp->ppe.ppeds_comp_ring.ppe_wbm2sw_ring);
 }
 
-- 
2.34.1

